File: 2/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *dummy = new ListNode(0);
        ListNode *firstItem = dummy;

        int remainingSum = 0;
        
        while(l1 != NULL || l2 != NULL) {
            int sum = remainingSum;
            if (l1 != NULL) sum += l1->val;
            if (l2 != NULL) sum += l2->val;
            remainingSum = sum/10;
            dummy->val = sum%10;
            
            if (l1 != NULL) l1 = l1->next;
            if (l2 != NULL) l2 = l2->next;
            
            if (l1 == NULL && l2 == NULL) {
                dummy->next = NULL;
            } else {
                dummy->next = new ListNode(0);
                dummy = dummy->next;
            }
        }
        
        if (remainingSum > 0) {
            dummy->next = new ListNode(remainingSum);
        }

        return firstItem;
    }
};

File: 2/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := new(ListNode)
	firstItem := dummy
	carry := 0

	for l1 != nil || l2 != nil {
		sum := carry
		if l1 != nil {
			sum += l1.Val
		}
		if l2 != nil {
			sum += l2.Val
		}
		carry = sum / 10
		dummy.Val = sum % 10

		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}

		if l1 == nil && l2 == nil {
			dummy.Next = nil
		} else {
			dummy.Next = new(ListNode)
			dummy = dummy.Next
		}
	}

	if carry > 0 {
		dummy.Next = new(ListNode)
		dummy.Next.Val = carry
	}
	return firstItem
}